include/rocksdb/flush_manager.h

```diff
@@ -0,0 +1,36 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <functional>
+#include <vector>
+
+namespace rocksdb {
+
+class ExternalFlushManager {
+ public:
+  virtual ~ExternalFlushManager() {}
+
+  std::function<void(uint32_t, std::vector<std::vector<uint32_t>>*)>
+      OnManualFlush;
+
+  std::function<void(std::vector<std::vector<uint32_t>>*)>
+      OnHandleWriteBufferFull;
+
+  std::function<void(std::vector<std::vector<uint32_t>>*)> OnSwitchWAL;
+
+  std::function<void(const std::vector<uint32_t>&,
+                     std::vector<std::vector<uint32_t>>*)>
+      OnScheduleFlushes;
```

Comment:
Somehow I have a difficulty in imagine how users are supposed to keep the context per DB. If they are doing things like CompactionFilter, or TableFactory, the context can be stored in class itself. Now how do they supposed to keep some context?

Reply:
They should inherit `ExternalFlushMananger` class and keep the context there.

Reply:
No, this is the only reason for this design.

Reply:
Can we still use normal function, but leaves the default to be the current behavior if they don't override it?